### PR TITLE
feat(forge/test): testrunner block summary

### LINKED ISSUE #1004: Add report after each test run
### Component

Forge

### Describe the feature you would like

rn the output of a test runner is like this

```console
Running x tests for <A>.<A>
[PASS] .....

Running x tests for <B>.<B>
[PASS] .....
```

It'd be useful to wrap up each block with an overview

```console
test result: ok. X passed; Y failed; Z ignored; finished in 19.04s
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ecmendenhall**
## Motivation

Resolves #1004 

Print a summary including pass/fail count and duration after each test suite block.

forge suite summary

## Solution
- `runner.rs`: Refactor map of test names/results to a `SuiteResult` struct that stores run duration and results.
- `test.rs`: Format and print summary following each suite run.

## Notes
- I am pretty new to Rust, so all comments are very welcome. This is mostly compiler-driven development.
- No ignored count yet.

### CODE REVIEW

File: cli/src/cmd/forge/test.rs [hunk 1]
```diff
@@ -288,14 +290,11 @@ pub struct TestOutcome {
     /// Whether failures are allowed
     allow_failure: bool,
     /// All test results `contract -> (test name -> TestResult)`
```
Comment (mattsse) [inline cli/src/cmd/forge/test.rs:292]: revise docs

File: cli/src/cmd/forge/test.rs [hunk 2]
```diff
@@ -346,6 +348,21 @@ impl TestOutcome {
     }
 }
 
+fn block_summary(duration: Duration, outcome: TestOutcome) {
```
Comment (mattsse) [inline cli/src/cmd/forge/test.rs:351]: can we make this a function of `TestOutcome` instead: `TestOutcome::summary -> String`?

File: cli/src/cmd/forge/test.rs [hunk 3]
```diff
@@ -346,6 +348,21 @@ impl TestOutcome {
...
+    let failed = outcome.failures().count();
+    let result = match failed {
+        0 => Colour::Green.paint("[ OK ]"),
+        _ => Colour::Red.paint("[FAIL]"),
+    };
```
Comment (mattsse) [inline cli/src/cmd/forge/test.rs:356]: more idiomatic to do an `let result = if outcome.failures().count() == 0 {} else {};`here

File: forge/src/runner.rs [hunk 4]
```diff
@@ -13,7 +13,31 @@ use foundry_evm::{
 use proptest::test_runner::TestRunner;
 use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
 use serde::{Deserialize, Serialize};
-use std::{collections::BTreeMap, fmt, time::Instant};
+use std::{
+    collections::BTreeMap,
+    fmt,
+    time::{Duration, Instant},
+};
+
```
Comment (mattsse) [inline forge/src/runner.rs:21]: please some basic docs, that this represents a set of tests and how long they took

File: forge/src/runner.rs [hunk 5]
```diff
@@ -233,27 +257,30 @@ impl<'a, DB: DatabaseRef + Send + Sync> ContractRunner<'a, DB> {
...
-                    counterexample: None,
-                    logs: setup.logs,
-                    kind: TestKind::Standard(0),
-                    traces: setup.traces,
-                    labeled_addresses: setup.labeled_addresses,
-                },
-            )]
-            .into())
+            return Ok(SuiteResult::new(
+                Instant::now().duration_since(start),
```
Comment (mattsse) [inline forge/src/runner.rs:269]: can use `start.elapsed()` instead

File: forge/src/runner.rs [hunk 6]
```diff
@@ -280,16 +307,17 @@ impl<'a, DB: DatabaseRef + Send + Sync> ContractRunner<'a, DB> {
             })
             .collect::<Result<BTreeMap<_, _>>>()?;
 
+        let duration = Instant::now().duration_since(start);
```
Comment (mattsse) [inline forge/src/runner.rs:310]: use elapsed instead

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for this!
overall this looks pretty good already,

I few smol nits though

**Review by mattsse — APPROVED** [review]
lgtm!
cc @gakonst

File: cli/src/cmd/forge/test.rs (hunk 2 above)
Comment (ecmendenhall) [reply cli/src/cmd/forge/test.rs]: Sure—are there any Rust specific rules of thumb on when to do this vs. use free functions? The only slightly tricky thing here is the `duration`, which is not defined for a `TestOutcome` with multiple results.

I left `duration` as an external argument to this function, but we _could_ make the `duration` the sum of all the `SuiteResult` durations for a `TestOutcome`, if it is OK to sum up multiple `Duration`s in that way. What do you think?

File: cli/src/cmd/forge/test.rs [hunk 7]
```diff
@@ -351,6 +356,18 @@ impl TestOutcome {
         }
         Ok(())
     }
+
+    pub fn summary(&self, duration: Duration) -> String {
+        let failed = self.failures().count();
+        let result = if failed == 0 { "ok" } else { "fail" };
+        format!(
+            "Test result: {}. {} passed; {} failed; finished in {:.2?}",
+            result,
```
Comment (onbjerg) [inline cli/src/cmd/forge/test.rs:365]: nit: can we color ok/fail?

**Review by onbjerg — APPROVED** [review]
One very very small nit, but LGTM :smile: Great work!

**Comment by ecmendenhall** [comment]
Colored the ok/fail and moved `duration` according to [this comment](https://github.com/gakonst/foundry/pull/1059#discussion_r836500479).

**Review by gakonst — APPROVED** [review]
love it

**Pull request merged**